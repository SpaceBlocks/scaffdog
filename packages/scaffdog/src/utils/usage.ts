import stripAnsi from 'strip-ansi';
import chalk from 'chalk';

const length = (s: string) => stripAnsi(s).length;

export type UsageSectionContent = {
  name: string;
  summary: string;
};

export type UsageSection = {
  header?: string;
  content: string | UsageSectionContent[];
  raw?: boolean;
};

export const buildUsage = (sections: UsageSection[]): string => {
  const maxContentSize = sections.reduce((max, section) => {
    if (!Array.isArray(section.content)) {
      return max;
    }

    return section.content.reduce((acc, cur) => {
      return Math.max(acc, length(cur.name));
    }, max);
  }, 0);

  return sections
    .map((section) => {
      const wrapContent = (s: string) => (section.raw === true ? s : `  ${s}`);
      const line: string[] = [];

      if (section.header != null) {
        line.push(chalk.bold(section.header));
      }

      if (typeof section.content === 'string') {
        line.push(wrapContent(section.content));
      } else {
        section.content.forEach((content) => {
          const padding = ' '.repeat(maxContentSize - length(content.name));

          line.push(
            wrapContent(`${content.name}${padding}  ${content.summary}`),
          );
        });
      }

      return line.join('\n');
    })
    .join('\n\n');
};
