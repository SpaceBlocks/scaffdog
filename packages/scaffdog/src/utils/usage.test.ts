import stripAnsi from 'strip-ansi';
import { describe, expect, test } from 'vitest';
import { buildUsage } from './usage.js';

describe('buildUsage', () => {
  test('empty', () => {
    const output = buildUsage([]);

    expect(output).toBe('');
  });

  test('sections', () => {
    const output = buildUsage([
      {
        content: 'This is simple content.',
      },
      {
        content: 'Raw content\nOne\nTwo',
        raw: true,
      },
      {
        header: 'Title line',
        content: 'This is simple content.',
      },
      {
        content: [
          {
            name: '-f, --force',
            summary: 'Summary 1',
          },
          {
            name: '    --flag',
            summary: 'Summary 2',
          },
        ],
      },
    ]);

    expect(stripAnsi(output)).toMatchSnapshot();
  });
});
