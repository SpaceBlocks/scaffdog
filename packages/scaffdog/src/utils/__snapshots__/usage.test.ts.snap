// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`buildUsage > sections 1`] = `
"  This is simple content.

Raw content
One
Two

Title line
  This is simple content.

  -f, --force  Summary 1
      --flag   Summary 2"
`;
